File: Leet-153. Find Minimum in Rotated Sorted Array/solution-logn.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findMin = function(nums) {
    if (nums === null || nums.length === 0) {
        return Number.MAX_SAFE_INTEGER;
    }
    
    var min = nums[0];
    var left = 0;
    var right = nums.length - 1;
    
    while (left <= right) {
        var mid = left + parseInt((right - left) / 2);
        
        if (nums[mid] > nums[left]) {
            min = Math.min(min, nums[left]);
            left = mid + 1;
        } else if (nums[mid] < nums[left]) {
            min = Math.min(min, nums[mid]);
            right = mid - 1;
        } else {                                    // when left === right || right - left < 2
            min = Math.min(nums[left], min);
            left++;
        }
    }
    
    return min;
};

File: Lint-458. Last Position of Target/solution-logn.java
public class Solution {
    /**
     * @param nums: An integer array sorted in ascending order
     * @param target: An integer
     * @return: An integer
     */
    public int lastPosition(int[] nums, int target) {
        if (nums == null || nums.length == 0 || target < nums[0] || target > nums[nums.length - 1]) {
            return -1;
        }
        
        int left = 0, right = nums.length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            
            if (nums[mid] == target) {
                if (mid == nums.length - 1) {
                    return mid;
                } else if (nums[mid + 1] == target) { 
                    left = mid + 1;
                } else {
                    return mid;
                }
            } else if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        
        return -1;
    }
}